File: problems/3001/solution.go
func minMovesToCaptureTheQueen(a int, b int, c int, d int, e int, f int) int {
    between := func(x, y, z int) bool {
        if x > z { x, z = z, x }
        return x < y && y < z
    }
    abs := func(x int) int { if x < 0 { return -x }; return x }

    rookBlocked := (a == e && c == a && between(b, d, f)) ||
        (b == f && d == b && between(a, c, e))
    if (a == e || b == f) && !rookBlocked { return 1 }

    bishopAttacks := abs(c-e) == abs(d-f)
    bishopBlocked := bishopAttacks && abs(a-e) == abs(b-f) &&
        between(c, a, e) && between(d, b, f)
    if bishopAttacks && !bishopBlocked { return 1 }
    return 2
}

File: problems/3001/solution.rs
impl Solution {
    pub fn min_moves_to_capture_the_queen(a: i32, b: i32, c: i32, d: i32, e: i32, f: i32) -> i32 {
        let between = |x: i32, y: i32, z: i32| x.min(z) < y && y < x.max(z);
        let rook_blocked = (a == e && c == a && between(b, d, f)) ||
                           (b == f && d == b && between(a, c, e));
        if (a == e || b == f) && !rook_blocked { return 1; }

        let bishop_attacks = (c - e).abs() == (d - f).abs();
        let bishop_blocked = bishop_attacks && (a - e).abs() == (b - f).abs() &&
                             between(c, a, e) && between(d, b, f);
        if bishop_attacks && !bishop_blocked { 1 } else { 2 }
    }
}
